### PR TITLE
dell-dock: Add support for flashing Thunderbolt over I2C

### PR DESCRIPTION
**Pull request by superm1**
This uses an API in the Realtek USB 3.1G2 controller to perform
the flash procedure.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by superm1** [comment]
@YehezkelShB, just FYI for you too if you want to take a look

File: plugins/dell-dock/fu-dell-dock-hid.c [hunk 1]
```diff
@@ -374,3 +394,131 @@ fu_dell_dock_hid_i2c_read (FuDevice *self,
...
+	return TRUE;
+}
+
+gboolean
+fu_dell_dock_hid_tbt_authenticate (FuDevice *self,
+				   const FuHIDI2CParameters *parameters,
+				   GError **error)
+{
+	FuTbtCmdBuffer cmd_buffer = {
+	    .cmd = HUB_CMD_READ_DATA, /* It's a special write command that reads status result */
```
Comment (hughsie) [inline plugins/dell-dock/fu-dell-dock-hid.c]: maybe `/* special write command...`

File: plugins/dell-dock/fu-dell-dock-i2c-ec.c [hunk 2]
```diff
@@ -524,6 +535,12 @@ fu_dell_dock_ec_modify_lock (FuDevice *device,
 		 fu_device_get_name (device),
 		 fu_device_get_id (device));
 
+	if (unlocked)
+		self->dock_unlock_status |= 1 << target;
+	else
+		self->dock_unlock_status &= ~(1 << target);
+	g_debug ("Current overall unlock status: 0x%08x", self->dock_unlock_status);
```
Comment (hughsie) [inline plugins/dell-dock/fu-dell-dock-i2c-ec.c]: lowercase debugs pls

File: plugins/dell-dock/fu-dell-dock-i2c-tbt.c [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+						 i,
+						 buffer,
+						 write_size,
+						 &tbt_base_settings,
+						 error))
+			return FALSE;
+
+		fu_device_set_progress_full (device, i, image_size);
+	}
+	g_debug ("Writing took %f seconds",
```
Comment (hughsie) [inline plugins/dell-dock/fu-dell-dock-i2c-tbt.c]: same

File: plugins/dell-dock/fu-dell-dock-i2c-tbt.c [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+		 g_timer_elapsed (timer, NULL));
+
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+
+	if (fu_device_has_custom_flag (device, "skip-restart")) {
+		g_debug ("Skipping TBT reset per quirk request");
```
Comment (hughsie) [inline plugins/dell-dock/fu-dell-dock-i2c-tbt.c]: same

File: plugins/dell-dock/fu-plugin-dell-dock.c [hunk 5]
```diff
@@ -68,6 +68,19 @@ fu_plugin_dell_dock_probe (FuPlugin *plugin,
 					      error))
 		return FALSE;
 
+	/* create TBT endpoint
```
Comment (hughsie) [inline plugins/dell-dock/fu-plugin-dell-dock.c]: comment on one line?

File: plugins/dell-dock/fu-dell-dock-i2c-tbt.c [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+		g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_INVALID_FILE,
+			     "Image is not intended for this system (0x%x)",
+			     target_system);
+		return FALSE;
+	}
+
+	buffer += start_offset;
+	image_size -= start_offset;
+
+	g_debug ("Waking Thunderbolt controller");
```
Comment (hughsie) [inline plugins/dell-dock/fu-dell-dock-i2c-tbt.c]: same

File: plugins/dell-dock/fu-dell-dock-i2c-ec.c [hunk 7]
```diff
@@ -524,6 +535,12 @@ fu_dell_dock_ec_modify_lock (FuDevice *device,
 		 fu_device_get_name (device),
 		 fu_device_get_id (device));
 
+	if (unlocked)
+		self->dock_unlock_status |= 1 << target;
+	else
+		self->dock_unlock_status &= ~(1 << target);
```
Comment (hughsie) [inline plugins/dell-dock/fu-dell-dock-i2c-ec.c]: I know what you're doing, but is there any BIT_SET and BIT_CLEAR macros we can use?

File: plugins/dell-dock/fu-dell-dock-hid.c [hunk 8]
```diff
@@ -374,3 +394,131 @@ fu_dell_dock_hid_i2c_read (FuDevice *self,
...
+		if (result == 0)
+			break;
+		g_debug ("attempt %d/%d: Thunderbolt authenticate failed: %x",
+			  i, TBT_MAX_RETRIES, result);
+		g_usleep (500000);
+	}
+	if (result != 0) {
+		g_set_error (error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL,
+			     "Thunderbolt authentication failed with status code 0x%x",
+			     result);
```
Comment (hughsie) [inline plugins/dell-dock/fu-dell-dock-hid.c]: can we map `result` to a proper error message?

File: plugins/dell-dock/fu-dell-dock-i2c-ec.c (hunk 7 above)
Comment (superm1) [reply plugins/dell-dock/fu-dell-dock-i2c-ec.c]: How about locally defining these in the file:
```
#define BIT(x)                          (1 << x)
#define BIT_SET(x,y)		(x |= (y))
#define BIT_CLEAR(x,y)		(x &= (~y))
```

And then it turns into
```
if (unlocked)
        BIT_SET (self->dock_unlock_status, BIT(target));
else
        BIT_CLEAR (self->dock_ulock_status, BIT(target));
```

Comment (superm1) [reply plugins/dell-dock/fu-dell-dock-i2c-ec.c]: I've added that now.

Comment (hughsie) [reply plugins/dell-dock/fu-dell-dock-i2c-ec.c]: Isn't

    #define BIT_SET(x,y)			(x |= (1<<(y)))
    #define BIT_CLEAR(x,y)			(x &= (~1<<(y)))

the usual form?

Comment (superm1) [reply plugins/dell-dock/fu-dell-dock-i2c-ec.c]: That way is fine to me too, just adjusted.

**Comment by superm1** [comment]
Actually I think there was a order of parenthesis error in CI from your `BIT_CLEAR` proposal leading to a left shift error.
I'll make sure the fixup builds on all the CI and then push it to master.

**Pull request merged**